File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.cc
#include <bits/stdc++.h>
using namespace std;

int parent[10002];

int find(int x);
void Union(int x, int y);

int main() {
    int v, e, a, b, c;
    vector<pair<int, pair<int, int>>> tree;

    cin >> v >> e;
    for(int i = 0; i < e; i++) {
        cin >> a >> b >> c;
        tree.push_back({c, {a, b}});
    }
    sort(tree.begin(), tree.end());

    for(int i = 0; i <= v; i++) 
        parent[i] = i;

    int sum = 0;
    for(int i = 0; i < e; i++) {
        int k = tree[i].first;
        int m = tree[i].second.first;
        int n = tree[i].second.second;
        if(find(m) == find(n)) continue; // If cycle pass

        Union(m, n);
        sum += k;
    }
    
    cout << sum << "\n";
    return 0;
}

int find(int x) 
{
    if(parent[x] == x) return x;
    else return find(parent[x]);
}

void Union(int x, int y) 
{
    int xRoot = find(x);
    int yRoot = find(y);
    parent[yRoot] = xRoot;
}

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
class Set:
    def __init__(self):
        self.p = list(range(V+1))
        self.rank = [1] * (V+1)

    def find(self, x):
        if self.p[x] != x:
            self.p[x] = self.find(self.p[x])
        return self.p[x]

    def union(self, x, y):
        px = self.find(x)
        py = self.find(y)

        if px == py:
            return False

        if self.rank[px] > self.rank[py]:
            self.p[py] = px
        elif self.rank[px] < self.rank[py]:
            self.p[px] = py
        else:
            self.p[py] = px
            self.rank[px] += 1
        return True

def kruskal():
    mst =[]
    U = Set()

    edges.sort(key=lambda x:x[2])
    for edge in edges:
        s, e, w = edge
        if U.union(s, e):
            mst.append(w)

    return mst

V, E = map(int, input().split())
edges = [list(map(int, input().split())) for _ in range(E)]

mst = kruskal()
print(sum(mst))